numpyro/distributions/censored.py

```diff
@@ -0,0 +1,251 @@
+
+                    log f(value)    if censored == 0
+                    log (1 - F(value))    if censored == 1
+
+            where f is the density and F the cumulative distribution function of ``base_dist``.
+
+            This is commonly used in survival analysis, where event times are positive,
+            but the approach is more general and can be applied to any distribution
+            with a cumulative distribution function, regardless of support.
+
+            In R's ``survival`` package notation, this corresponds to
+            ``Surv(time, event, type = 'right')``.
+
+            Example:
+
+                    Surv(time = c(5, 8, 10), event = c(1, 0, 1))
+
+            means:
+
+                    subject 1 had an event at t=5
+                    subject 2 was censored at t=8
+                    subject 3 had an event at t=10
+
+    **Example:**
+
+    .. doctest::
+
+            >>> from jax import numpy as jnp
+            >>> from numpyro import distributions as dist
+            >>> base = dist.Exponential(rate=0.1)
+            >>> surv_dist = dist.RightCensoredDistribution(base, censored=jnp.array([0, 1, 0]))
+            >>> loglik = surv_dist.log_prob(jnp.array([5., 8., 10.]))
+    """
+
+    arg_constraints = {"censored": constraints.boolean}
+    pytree_data_fields = ("base_dist", "censored", "_support")
+
+    def __init__(
+        self,
+        base_dist: DistributionT,
+        censored: ArrayLike = False,
+        *,
+        validate_args: Optional[bool] = None,
+    ):
+        # test if base_dist has an implemented cdf method
+        if not hasattr(base_dist, "cdf"):
+            raise TypeError(
+                f"{type(base_dist).__name__} does not have a 'cdf' method. "
+                "Censored distributions require a base distribution with an "
+                "implemented cumulative distribution function."
+            )
+
+        # Optionally test that cdf actually works (in validate_args mode)
+        if validate_args:
+            try:
+                test_val = base_dist.support.feasible_like(jnp.array(0.0))
+                _ = base_dist.cdf(test_val)
+            except (NotImplementedError, AttributeError) as e:
+                raise TypeError(
+                    f"{type(base_dist).__name__}.cdf() is not properly implemented."
+                ) from e
+        batch_shape = lax.broadcast_shapes(base_dist.batch_shape, jnp.shape(censored))
+        self.base_dist = jax.tree.map(
+            lambda p: promote_shapes(p, shape=batch_shape)[0], base_dist
+        )
+        self.censored = jnp.array(
+            promote_shapes(censored, shape=batch_shape)[0], dtype=jnp.bool
+        )
+        self._support = base_dist.support
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    def sample(
+        self, key: Optional[jax.dtypes.prng_key], sample_shape: tuple[int, ...] = ()
+    ) -> ArrayLike:
+        return self.base_dist.sample(key, sample_shape)
```

Comment:
You'll need self.base_dict.expand(self.batch_shape).sample(...). promote_shapes does not broadcast the parameters, it only adds some missing singleton dimensions.

Reply:
good catch; updated in newest commit and also added a test